README.md
# Rust Server (Rcon) Panel

RustServerPanel Is a server admin panel for the game Rust using RCON (A websocket communication) to administrat it.

Actually is in development and you can see the progress below.

ℹ If you want to see the full changelog [Click here](https://github.com/iSnakeBuzz/RustRconPanel/blob/master/CHANGELOG.md "CHANGELOG.MD").

#### Work in progress screenshots.

- Added the console in the Server section.
  !['Console'](https://i.imgur.com/G7rVzB0.png)

- Changed background color for one more nicer. Aslo added a new window to add the new servers.
  !['Changed color'](https://i.imgur.com/5EBMNZT.png)
  !['New window'](https://i.imgur.com/9DxBHnR.png)

- Working on design (Inspired by [Discord](http://discord.com/))
  !['First Screenshot'](https://i.imgur.com/KF1U9fb.png)

#### Used technologies

- [ElectronJS](https://electronjs.org/)
- [ReactJS](https://create-react-app.dev/)


CHANGELOG.md
# 17/01/2021

- Started server page.
- Update server list when a new one is created.
- Handle when a servers has been added.
- I've installed Moment JS for a more fluent management of Dates.
- Added a custom Scrollbar.
- Added server action buttons.
- Added a terminal (In test mode)
- Started the connections with Websocket
- Added functionalities to console

# 14/01/2021

- Changed background color
- Changed how to build the ReactJS in the production build.
- Added a new window that make easy to add a new server.
- Changed some styles in the main material UI theme.
- Working on how store the data.


electronApp.js
const { app, BrowserWindow, ipcMain } = require('electron');
const path = require('path');
const isDev = require('electron-is-dev');


let mainWindow;

/* Create main Window */
const createWindow = () => {
    /* Creating the Main Window */
    mainWindow = new BrowserWindow({
        minWidth: 800,
        minHeight: 450,
        width: 940,
        height: 800,
        darkTheme: true,
        center: true,
        frame: false,
        webPreferences: { nodeIntegration: true }
    });

    /* Loading website. If dev, loads a local URL, otherwise runs from build folder. */
    mainWindow.loadURL(isDev ? 'http://localhost:3000' : `file://${path.join(__dirname, './build/index.html')}`);

    /* Enable dev tools if isDev is true. (CTRL + SHIFT + I opens the dev tools manually) */
    // if (isDev) mainWindow.webContents.openDevTools();

    mainWindow.on('closed', () => mainWindow = null);
}

app.on('ready', createWindow);

app.on('window-all-closed', () => {
    if (process.platform !== 'darwin') {
        app.quit();
    }
});

app.on('activate', () => {
    if (mainWindow === null) {
        createWindow();
    }
});

/* IPC communicatino with FrontEnd */
ipcMain.on('minimize', (e, arg) => {
    mainWindow.minimize();

    /* I return a value to prevent bugs :) */
    e.returnValue = "pong";
});

ipcMain.on('maximize', (e, arg) => {
    if (!mainWindow.isMaximized()) mainWindow.maximize();
    else mainWindow.unmaximize();

    e.returnValue = "pong";
});

ipcMain.on('close', (e, arg) => {
    app.quit();
    e.returnValue = "pong";
});

public/electron/ipcmc.js
const { app, BrowserWindow, ipcMain } = require('electron');
const serve = require('electron-serve');
const isDev = require('electron-is-dev');

/* Add Server Window modal */
let ADD_SERVER_WINDOW;

/* IPC communicatino with FrontEnd */
ipcMain.on('add_server', (e, arg) => {
    /* Creating a window to add the server */
    if (!ADD_SERVER_WINDOW) {
        ADD_SERVER_WINDOW = new BrowserWindow({
            width: 250,
            height: 450,

            darkTheme: true,
            center: true,
            frame: false,
            resizable: false,
            webPreferences: { nodeIntegration: true }
        });
        ADD_SERVER_WINDOW.loadURL(isDev ? 'http://localhost:3000/add-server' : 'app://-/add-server')
        e.returnValue = true;

        /* Handling window close */
        ADD_SERVER_WINDOW.on('close', () => {
            ADD_SERVER_WINDOW = undefined;
            updateServers();
        })
    } else e.returnValue = false;
});

const updateServers = () => {
    let windows = BrowserWindow.getAllWindows();
    mainWindow = windows[windows.length - 1];

    console.log("Updating servers.", mainWindow)
    mainWindow.webContents.postMessage('update_servers', 'ping');
}

src/Connections/RconService.js
class RconService {

    constructor(address, port, password) {
        this.urlString = `ws://${address}:${port}/${password}`;
        this.socket = new WebSocket(this.urlString);
        this.identifier = 1001;
        this.callbacks = {};

        this.socket.addEventListener('open', (e) => console.log(`[Socket] Connected to ${this.urlString}`));
        this.socket.addEventListener('close', this.disconnect);
        this.socket.addEventListener('error', (e, error) => console.log(`[Socket] Error at ${this.urlString}`, e, error));

        this.socket.addEventListener('message', (e) => {
            console.log(`[Socket] Receiving data from ${this.urlString}`)
            let data = JSON.parse(e.data);
            let dataIdentifier = data.Identifier;

            if (dataIdentifier > 1000) {
                let item = this.callbacks[dataIdentifier];
                if (item != null) {
                    item.callback(data);
                }

                this.callbacks[dataIdentifier] = null;
                return;
            }
        })
    }

    onMessage(callback) {
        this.socket.addEventListener('message', callback);
    }

    onOpen(callback) {
        this.socket.addEventListener('open', callback);
    }

    sendMessage(message, identifier) {
        let packet = {
            Identifier: identifier,
            Message: message,
            Name: "WebRcon"
        };

        this.socket.send(JSON.stringify(packet));
    }

    request(message, callback) {
        if (!this.isReady) return;
        this.identifier++;

        this.callbacks[this.identifier] = {
            callback: callback
        };
        this.sendMessage(message, this.identifier);
    }

    disconnect() {
        if (this.socket) {
            this.socket.close();
            this.socket = null;
        }

        this.callbacks = {};
    }

    isReady() {
        return this.socket.readyState === 1;
    }

}

export default RconService;

public/electron.js
const { app, BrowserWindow, ipcMain } = require('electron');
const path = require('path');
const isDev = require('electron-is-dev');
const serve = require('electron-serve');
const serveReact = serve({ directory: path.join(__dirname, '../build') });
require('./electron/ipcmc')

let mainWindow;

/* Create main Window */
const createWindow = async () => {
    /* Creating the Main Window */
    mainWindow = new BrowserWindow({
        minWidth: 940,
        minHeight: 500,
        width: 940,
        height: 500,
        darkTheme: true,
        center: true,
        frame: false,
        webPreferences: { nodeIntegration: true }
    });

    /* Loading website. If dev, loads a local URL, otherwise runs from build folder. */
    /* Using electron-serve on production to serve the single-page web app  */
    if (isDev) mainWindow.loadURL('http://localhost:3000');
    else serveReact(mainWindow);

    /* Enable dev tools if isDev is true. (CTRL + SHIFT + I opens the dev tools manually) */
    // if (isDev) mainWindow.webContents.openDevTools();

    mainWindow.on('closed', () => mainWindow = null);

}

app.whenReady().then(() => createWindow());

app.on('window-all-closed', () => {
    if (process.platform !== 'darwin') {
        app.quit();
    }
});

app.on('activate', () => {
    if (mainWindow === null) createWindow();
});

/* IPC communicatino with FrontEnd */
ipcMain.on('minimize', (e, arg) => {
    let focusedWindow = BrowserWindow.getFocusedWindow();
    focusedWindow.minimize();

    /* I return a value to prevent bugs :) */
    e.returnValue = "pong";
});

ipcMain.on('maximize', (e, arg) => {

    let focusedWindow = BrowserWindow.getFocusedWindow();

    if (!focusedWindow.isMaximized()) focusedWindow.maximize();
    else focusedWindow.unmaximize();

    e.returnValue = "pong";
});

ipcMain.on('close', (e, arg) => {
    let focusedWindow = BrowserWindow.getFocusedWindow();

    if (focusedWindow.id === mainWindow.id) app.quit();
    else focusedWindow.close();

    e.returnValue = "pong";
});